frontend/packages/core/src/navigation.tsx

```diff
@@ -0,0 +1,171 @@
+ * setting search params. See https://github.com/ReactTraining/react-router/issues/7743#issuecomment-785435977
+ * for additional context.
+ */
+interface CustomPartialPath extends Pick<PartialPath, "pathname" | "hash"> {
+  /**
+   * Search params to include in navigation.
+   */
+  search?: URLSearchParamsInit;
+}
+
+type To = string | CustomPartialPath;
+
+/**
+ * React Router specific navigation options.
+ */
+interface ReactRouterNavigateOptions {
+  replace?: boolean;
+  state?: State;
+}
+
+/**
+ * Custom navigation options to control Clutch specific routing functionality.
+ */
+export interface NavigateOptions extends ReactRouterNavigateOptions {
+  // Include current URL in state as origin. Defaults to false.
+  origin?: boolean;
+  // Persist UTM search params after navigation. Defaults to true.
+  utm?: boolean;
+}
+
+/**
+ * Convert custom search param object to string.
+ */
+const convertSearchParam = (params: URLSearchParams) => {
+  const p = params.toString();
+  if (p.indexOf("=") < 0) {
+    return "";
+  }
+  return `?${p}`;
+};
+
+/**
+ * Returns an imperative method for changing the location.
+ *
+ * This method wraps useNavigate from react-router but has custom
+ * logic that:
+ *   * Automatically preserves UTM parameters during navigation
+ *   * Easily preserves the origin of navigation in state
+ *   * Provides an easy interface to specify search parameters
+ *
+ * @see https://reactrouter.com/api/useNavigate for more information.
+ */
+const useNavigate = () => {
+  const navigation = rrUseNavigate();
+  const [currentSearchParams] = rrUseSearchParams();
+  const customNavigate = (to: To, options: NavigateOptions = {}) => {
+    const finalNavOptions = {} as ReactRouterNavigateOptions;
+    if (options?.replace) {
+      finalNavOptions.replace = options.replace;
+    }
+    if (options?.state) {
+      finalNavOptions.state = options.state;
+    }
+    let newSearchParams = {} as URLSearchParamsInit;
+    if (typeof to !== "string") {
+      newSearchParams = (to?.search || {}) as URLSearchParamsInit;
+    }
+    const searchParams = createSearchParams(newSearchParams);
+    if (options?.utm || options?.utm === undefined) {
+      UTM_PARAMS.forEach(p => {
+        const param = currentSearchParams.get(p);
+        if (param && !searchParams.get(p)) {
+          searchParams.set(p, param);
+        }
+      });
+    }
+
+    if (options?.origin) {
+      const origin = `${window.location.pathname}${window.location.search}`;
+      // n.b. if origin is specified in the options don't overwrite it.
+      // @ts-ignore
+      if (!options.state?.origin) {
+        // @ts-ignore
+        finalNavOptions.state = { ...finalNavOptions.state, origin };
+      }
+    }
+
+    let navPath = {} as PartialPath;
+    if (typeof to === "string") {
+      navPath.pathname = to;
+    } else {
+      navPath = { pathname: to?.pathname, hash: to?.hash };
+    }
+    const [path, search] = navPath.pathname.split("?");
```

Comment:
Is there any utility that would simplify parsing the URL rather than needing to split?

Reply:
There isn't an easy way to split this since it's just the path. If I wanted to use a utility I would need to rely on the window location to get the hostname.